code/utils/ps.py

```diff
@@ -0,0 +1,83 @@
+import numpy as np
+import numpy
+
+import tensorflow as tf
+
+
+def _initialize_pk(shape,boxsize,kmin,dk):
+    """
+       Helper function to initialize various (fixed) values for powerspectra... not differentiable!
+    """
+    I = np.eye(len(shape), dtype='int') * -2 + 1
+
+    W = np.empty(shape, dtype='f4')
+    W[...] = 2.0
+    W[..., 0] = 1.0
+    W[..., -1] = 1.0
+    kedges = np.arange(kmin, np.pi * 10 / 1 + dk/2, dk)
+
+    k = [np.fft.fftfreq(N, 1. / (N * 2 * np.pi / L))[:pkshape].reshape(kshape) for N, L, kshape, pkshape in zip(shape, boxsize, I, shape)]
+    kmag = sum(ki ** 2 for ki in k) ** 0.5
+
+    xsum = np.zeros(len(kedges) + 1)
+    Nsum = np.zeros(len(kedges) + 1)
+
+    dig = np.digitize(kmag.flat, kedges)
+
+    xsum.flat += np.bincount(dig, weights=(W * kmag).flat, minlength=xsum.size)
+    Nsum.flat += np.bincount(dig, weights=W.flat, minlength=xsum.size)
+    dig = tf.Variable(dig,dtype=tf.int32)
+    Nsum = tf.Variable(Nsum,dtype=tf.complex64)
```

Comment:
I don't think these really have to be tf.Variables, but things work if they are tf variables? *shrug*

Reply:
maybe I could spend like 5 minutes to make this whole part differentiable too in case someone ever wants to take the derivative with respect to box size :P